docs/design/chart_format.md

```diff
@@ -0,0 +1,664 @@
+
+### Directory Layout
+A chart is laid out as follows. The top level directory (represented by the placeholder ROOT) must be the name of the chart. For example, if the chart is named `nginx`, the ROOT directory must be named `nginx/`.
+
+```
+ROOT/
+    Chart.yaml
+    README.md
+    LICENSE
+    docs/
+        some.md
+    templates/
+        some.yaml
+        some.jinja
+        some.jinja.schema
+```
+
+Templates are stored in a separate directory for the following reasons:
+
+
+* This future-proofs the format: we can add other directories at the top level and deprecate `templates/`.
+* It allows authors to add other files (such as documentation) in a straightforward way that will not confuse definitions-aware tools.
+* It allows for the possibility that a definition may be embedded inside of project code.
+
+
+Charts must not assume that they can perform file system operations to load another chart or supporting resources directly via the filesystem, nor should they store any operational files outside of the chart directory. It should be possible to archive the chart directory (e.g. `tar -cf ROOT.tar ROOT/`) and not lose any necessary information.
+
+### Generators and Templates
+The contents inside of the `templates/` directory may contain directives that can be processed by an external processor. Such a processor may read these files and generate new content.
+
+
+A generator reads through the files in `templates/` and looks for special file headers that indicate what generation tool needs to be run (if any) on a file. One such implementation of a generation tool is a template renderer. In any one pass through the `templates/` directory, more than one generation tool may be invoked.
+
+
+#### The Generator Header
+A file that requires a generation tool must declare this on the first line of the file using one of the following syntaxes:
+
+```
+//helm:generate TOOL [ARGS]\n
+#helm:generate TOOL [ARGS]\n
+/*helm:generate TOOL [ARGS]*/
+```
+
+`TOOL` corresponds to a registered tool name (see [this proposal](https://github.com/kubernetes/kubernetes/pull/18215#issuecomment-168885511)). `[ARGS]` are zero or more space-separated strings that will be passed unmodified to the tool. The first two formats require a terminating line feed. The third format requires a terminating `*/` sequence.
+
+
+**Example:**
+
+```
+#helm:generate dm_template
+{% set PROPERTIES = properties or {} %}
+{% set WORKERS = PROPERTIES['workers'] or 2 %}
+resources:
+- name: spark-master
+ type: ReplicationController
+ properties:
+   kind: ReplicationController
+   apiVersion: v1
+   metadata:
+     name: spark-master-controller
+   spec:
+…
+```
+
+
+In the example above, the `#helm:generate dm_template` indicates that the remainder of the file contents should be processed using whatever handler is registered for handling `dm_template`.
+
+
+Generator tools may take arguments. No format is specified for arguments. For example, one might use UNIX-like flags.
```

Comment:
We should support environment variables. I don't see why we wouldn't and helm already does this.